pkg/gcp/resources.go

```diff
@@ -488,12 +575,150 @@ func (r *gcpGKE) fromResourceDecription(resourceDesc []byte) (*containerpb.Creat
 	return createClusterRequest, nil
 }
 
-// getInstanceUrl returns a fully qualified URL for an instance
-func getInstanceUrl(project, zone, instance string) string {
-	return computeUrlPrefix + fmt.Sprintf("projects/%s/zones/%s/instances/%s", project, zone, instance)
+// GCP private service connect
+type privateServiceHandler struct {
+	GCPResourceHandler
+	addressesClient   *compute.AddressesClient
+	forwardingClient  *compute.ForwardingRulesClient
+	attachmentsClient *compute.ServiceAttachmentsClient
 }
 
-// getClusterUrl returns a fully qualified URL for a cluster
-func getClusterUrl(project, zone, cluster string) string {
-	return containerUrlPrefix + fmt.Sprintf("projects/%s/locations/%s/clusters/%s", project, zone, cluster)
+// Initialize necessary clients for the handler
+func (r *privateServiceHandler) initClients(ctx context.Context, clients *GCPClients) error {
+	addressesClient, err := clients.GetAddressesClient(ctx)
+	if err != nil {
+		return err
+	}
+	r.addressesClient = addressesClient
+
+	forwardingClient, err := clients.GetForwardingClient(ctx)
+	if err != nil {
+		return err
+	}
+	r.forwardingClient = forwardingClient
+
+	attachmentsClient, err := clients.GetServiceAttachmentsClient(ctx)
+	if err != nil {
+		return err
+	}
+	r.attachmentsClient = attachmentsClient
+
+	return nil
+}
+
+// Read and provision a private service connect endpoint to associate with a service attachment
+func (r *privateServiceHandler) readAndProvisionResource(ctx context.Context, resource *paragliderpb.CreateResourceRequest, subnetName string, resourceInfo *resourceInfo, additionalAddrSpaces []string) (string, string, error) {
+	description := &ServiceAttachmentDescription{}
+	err := json.Unmarshal(resource.Description, description)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to parse resource description: %w", err)
+	}
+	return r.createWithNetwork(ctx, *description, subnetName, resourceInfo)
+}
+
+// Get the resource information about a service attachment
+func (r *privateServiceHandler) getResourceInfo(ctx context.Context, resource *paragliderpb.CreateResourceRequest) (*resourceInfo, error) {
+	description := &ServiceAttachmentDescription{}
+	err := json.Unmarshal(resource.Description, description)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse resource description: %w", err)
+	}
+	urlParts := strings.Split(description.Url, "/")
+	serviceName := urlParts[len(urlParts)-1]
+	region := urlParts[len(urlParts)-3]
```

Comment:
Thoughts on using `parseUrl` in `plugin_utils.go`? Would be useful as some users might pass in full vs. partial url.